### PR TITLE
refactor BaseRequestBuilder

### PR DESCRIPTION
**Pull request by arthurvaverko**
- added a thread-safe unique request counter to help identify logs of request\response
- refactore the async post request handling
- added a new method ExecuteAsync which implements the async\ await feature

### CODE REVIEW

File: sources/csharp2/KalturaClient/Request/BaseRequestBuilder.cs [hunk 1]
```diff
@@ -43,18 +34,17 @@ public string Boundary
...
         abstract public MultiRequestBuilder Add(IRequestBuilder requestBuilder);
         abstract public object Deserialize(XmlElement xmlElement);
 
-        private void Log(string msg)
+        private void Log(string msg, string requestId)
         {
             if (client != null && client.Configuration.Logger != null)
             {
-                client.Configuration.Logger.Log(msg);
+                var msgtoLog = $"KalturaClient > [{client.ClientTag}] > [{requestId}] > {msg}";
```
Comment (tan-tan-kanarek) [inline sources/csharp2/KalturaClient/Request/BaseRequestBuilder.cs]: Please remove "KalturaClient and clientTag, they're always the same and if someone wants them he should add them in his own logger.

File: sources/csharp2/KalturaClient/Request/BaseRequestBuilder.cs [hunk 2]
```diff
@@ -75,231 +65,231 @@ public BaseRequestBuilder<T> SetCompletion(OnErrorHandler errorHandler)
...
+            if (this.client == null) { this.Build(client); }
+            if (this.client == null) { OnComplete(null, new ClientException("No client instance defined")); }
 
-            request.Headers = getHeaders();
-            request.AutomaticDecompression = DecompressionMethods.GZip | DecompressionMethods.Deflate;
-            request.Accept = "application/xml";
-            request.ContentType = getContentType();
+            var url = client.Configuration.ServiceUrl + "/api_v3" + getPath();
+
+            var requestId = (Interlocked.Increment(ref Client.REQUEST_COUNTER)).ToString("X5");
```
Comment (tan-tan-kanarek) [inline sources/csharp2/KalturaClient/Request/BaseRequestBuilder.cs]: requestId could be initiated in the constructor.

File: sources/csharp2/KalturaClient/Request/BaseRequestBuilder.cs [hunk 3]
```diff
@@ -43,18 +34,17 @@ public string Boundary
...
             this.service = service;
-            bufferRead = new byte[BUFFER_SIZE];
-            requestData = new StringBuilder(System.String.Empty);
         }
 
         abstract public MultiRequestBuilder Add(IRequestBuilder requestBuilder);
         abstract public object Deserialize(XmlElement xmlElement);
 
-        private void Log(string msg)
+        private void Log(string msg, string requestId)
```
Comment (tan-tan-kanarek) [inline sources/csharp2/KalturaClient/Request/BaseRequestBuilder.cs]: requestId could be a protected member, no need to pass it through methods.

File: sources/csharp2/KalturaClient/Request/BaseRequestBuilder.cs [hunk 4]
```diff
@@ -75,231 +65,231 @@ public BaseRequestBuilder<T> SetCompletion(OnErrorHandler errorHandler)
...
+        {
+            Params kparams = new Params();
+
+            if (client != null)
+                kparams.Add(client.RequestConfiguration.ToParams(false));
+
+            kparams.Add(base.ToParams(false));
+
+            if (includeServiceAndAction)
+                kparams.Add("service", service);
```
Comment (tan-tan-kanarek) [inline sources/csharp2/KalturaClient/Request/BaseRequestBuilder.cs]: I don't see that you included the action.

File: sources/csharp2/KalturaClient/Request/BaseRequestBuilder.cs [hunk 5]
```diff
@@ -75,231 +65,231 @@ public BaseRequestBuilder<T> SetCompletion(OnErrorHandler errorHandler)
             return this;
         }
 
+        public virtual Params getParameters(bool includeServiceAndAction)
```
Comment (tan-tan-kanarek) [inline sources/csharp2/KalturaClient/Request/BaseRequestBuilder.cs]: I don't see it's ever called with includeServiceAndAction = true

File: sources/csharp2/KalturaClient/Request/BaseRequestBuilder.cs [hunk 6]
```diff
@@ -75,231 +65,231 @@ public BaseRequestBuilder<T> SetCompletion(OnErrorHandler errorHandler)
...
-            return taskWrapper.Result;
+                if (t.Status == TaskStatus.Faulted) { OnComplete(null, t.Exception); return; }
+
+                var result = t.Result;
+                OnComplete(result, null);
+            }).ConfigureAwait(false);
         }
 
-        private void RequestCallback(IAsyncResult result)
+        public T ExecuteAndWaitForResponse(Client client = null)
```
Comment (tan-tan-kanarek) [inline sources/csharp2/KalturaClient/Request/BaseRequestBuilder.cs]: You can rename to ExecuteSync.

Comment (arthurvaverko) [reply sources/csharp2/KalturaClient/Request/BaseRequestBuilder.cs]: while i completely agree with you, I can't this was already there and will require all to change names .. backward comparability

File: sources/csharp2/KalturaClient/Request/BaseRequestBuilder.cs (hunk 4 above)
Comment (arthurvaverko) [reply sources/csharp2/KalturaClient/Request/BaseRequestBuilder.cs]: this method did not change from the original and was used as is .. 
it is a virtual method so i assume the concrete requests will override or add more logic to it.

please see original method: 
https://github.com/kaltura/clients-generator/blob/b201140892d8191aad4477b8a769ee273ba7cca2/sources/csharp2/KalturaClient/Request/BaseRequestBuilder.cs#L394

File: sources/csharp2/KalturaClient/Request/BaseRequestBuilder.cs (hunk 5 above)
Comment (arthurvaverko) [reply sources/csharp2/KalturaClient/Request/BaseRequestBuilder.cs]: this method did not change from the original and was used as is .. 
it is a virtual method so i assume the concrete requests will override or add more logic to it.

please see original method: 
https://github.com/kaltura/clients-generator/blob/b201140892d8191aad4477b8a769ee273ba7cca2/sources/csharp2/KalturaClient/Request/BaseRequestBuilder.cs#L394

File: sources/csharp2/KalturaClient/Request/BaseRequestBuilder.cs (hunk 1 above)
Comment (arthurvaverko) [reply sources/csharp2/KalturaClient/Request/BaseRequestBuilder.cs]: done: 4025cc3858f2210933ab5c6759e40afab0680c4c

File: sources/csharp2/KalturaClient/Request/BaseRequestBuilder.cs (hunk 2 above)
Comment (arthurvaverko) [reply sources/csharp2/KalturaClient/Request/BaseRequestBuilder.cs]: done: 4025cc3858f2210933ab5c6759e40afab0680c4c

File: sources/csharp2/KalturaClient/Request/BaseRequestBuilder.cs (hunk 3 above)
Comment (arthurvaverko) [reply sources/csharp2/KalturaClient/Request/BaseRequestBuilder.cs]: done: 4025cc3858f2210933ab5c6759e40afab0680c4c

**Pull request merged**